File: python/course-repo/exercise/03-dedup-via-id-reuse/src/webhooks/activities.py
import time

import httpx
from temporalio import activity

from .shared import WebhookDelivery


@activity.defn
def deliver_webhook(req: WebhookDelivery) -> int:
    activity.logger.info(
        "Delivering webhook for event %s to %s", req.event_id, req.url
    )
    headers = {"Idempotency-Key": activity.info().activity_id}
    response = httpx.post(req.url, json=req.payload, headers=headers, timeout=10.0)
    response.raise_for_status()
    # Tiny pause so the second duplicate-call lands while this activity is
    # still in-flight (id_conflict_policy applies; id_reuse_policy does not).
    time.sleep(1)
    return response.status_code

File: typescript/course-repo/solution/01-durable-job-queue/src/activities.ts
import { log } from '@temporalio/activity';
import type { WebhookDelivery } from './shared';

// Same activity whether this runs standalone or as a step inside a Workflow.
export async function deliverWebhook(req: WebhookDelivery): Promise<number> {
  log.info('Delivering webhook for event', { eventId: req.eventId, url: req.url });
  const response = await fetch(req.url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req.payload),
  });
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  return response.status;
}
